collector/drbd_linux.go

```diff
@@ -0,0 +1,211 @@
+			prometheus.GaugeValue,
+			1),
+		"ap": newDrbdNumericalMetric(
+			"application_pending",
+			"Number of block I/O requests forwarded to DRBD, but not yet answered by DRBD.",
+			prometheus.GaugeValue,
+			1),
+		"ep": newDrbdNumericalMetric(
+			"epochs",
+			"Number of Epochs currently on the fly.",
+			prometheus.GaugeValue,
+			1),
+		"oos": newDrbdNumericalMetric(
+			"out_of_sync_bytes",
+			"Amount of data known to be out of sync.",
+			prometheus.GaugeValue,
+			1024),
+	}
+	drbdStringPairMetrics = map[string]drbdStringPairMetric{
+		"ro": newDrbdStringPairMetric(
+			"node_role_is_primary",
+			"Whether the role of the node is in the primary state.",
+			"Primary"),
+		"ds": newDrbdStringPairMetric(
+			"disk_state_is_up_to_date",
+			"Whether the disk of the node is up to date.",
+			"UpToDate"),
+	}
+
+	drbdConnected = prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, "drbd", "connected"),
+		"Whether DRBD is connected to the partner.",
+		[]string{"device"}, nil)
+)
+
+type drbdCollector struct{}
+
+func init() {
+	Factories["drbd"] = NewDrbdCollector
+}
+
+func NewDrbdCollector() (Collector, error) {
+	return &drbdCollector{}, nil
+}
+
+func (c *drbdCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	statsFile := procFilePath("drbd")
+	f, err := os.Open(statsFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Debugf("Not collecting DRBD statistics, as %s does not exist: %s", statsFile)
+			return nil
+		}
+		return err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	scanner.Split(bufio.ScanWords)
+	device := "unknown"
+	for scanner.Scan() {
+		field := scanner.Text()
+		if kv := strings.Split(field, ":"); len(kv) == 2 {
+			if id, err := strconv.ParseUint(kv[0], 10, 64); err == nil && kv[1] == "" {
+				device = fmt.Sprintf("drbd%d", id)
+			} else if metric, ok := drbdNumericalMetrics[kv[0]]; ok {
+				// Numerical value.
+				value, _ := strconv.ParseFloat(kv[1], 64)
+				ch <- prometheus.MustNewConstMetric(
+					metric.desc, metric.valueType,
+					value*metric.multiplier, device)
+			} else if metric, ok := drbdStringPairMetrics[kv[0]]; ok {
+				// String pair value.
+				values := strings.Split(kv[1], "/")
+				ch <- prometheus.MustNewConstMetric(
+					metric.desc, prometheus.GaugeValue,
+					metric.isOkay(values[0]), device, "local")
+				ch <- prometheus.MustNewConstMetric(
+					metric.desc, prometheus.GaugeValue,
+					metric.isOkay(values[1]), device, "remote")
+			} else if kv[0] == "cs" {
+				// Connection state.
+				var connected float64 = 0
+				if kv[1] == "Connected" {
+					connected = 1
+				}
+				ch <- prometheus.MustNewConstMetric(
+					drbdConnected, prometheus.GaugeValue,
+					connected, device)
+
+			} else {
+				log.Infof("Don't know how to process key-value pair [%s: %s]", kv[0], kv[1])
+			}
+		} else {
+			log.Infof("Don't know how to process string %s", field)
+		}
+	}
+	return nil
```

Comment:
Please catch `scanner.Scan()` errors.  See https://github.com/prometheus/node_exporter/blob/master/collector/meminfo_numa_linux.go#L164-L181 for a good example.